### PR TITLE
[cxx-interop] Make `Cxx` Swift library static

### PR DESCRIPTION
**Pull request by egorzhdan**
Instead of a dynamic `swiftCxx.dylib` library, let's build a static library to simplify backdeployment and reduce potential compatibility difficulties in the future.

### CODE REVIEW

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

**Review by hyp — APPROVED** [review]

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

**Comment by egorzhdan** [comment]
Can't reproduce the macOS test failure locally 😕

```
******************** TEST 'Swift(macosx-x86_64) :: Interop/Cxx/stdlib/overlay/std-string-overlay.swift' FAILED ********************
Script:
--
: 'RUN: at line 1';   rm -rf "/Users/ec2-user/jenkins/workspace/swift-PR-macos-smoke-test/branch-main/build/buildbot_incremental/swift-macosx-x86_64/test-macosx-x86_64/Interop/Cxx/stdlib/overlay/Output/std-string-overlay.swift.tmp" && mkdir -p "/Users/ec2-user/jenkins/workspace/swift-PR-macos-smoke-test/branch-main/build/buildbot_incremental/swift-macosx-x86_64/test-macosx-x86_64/Interop/Cxx/stdlib/overlay/Output/std-string-overlay.swift.tmp" && xcrun --toolchain default --sdk '/Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.0.sdk' /Users/ec2-user/jenkins/workspace/swift-PR-macos-smoke-test/branch-main/build/buildbot_incremental/swift-macosx-x86_64/bin/swiftc -target x86_64-apple-macosx10.9  -module-cache-path /Users/ec2-user/jenkins/workspace/swift-PR-macos-smoke-test/branch-main/build/buildbot_incremental/swift-macosx-x86_64/swift-test-results/x86_64-apple-macosx10.9/clang-module-cache  -F '/Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/Library/Frameworks' -toolchain-stdlib-rpath -Xlinker -rpath -Xlinker '/Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/Library/Frameworks' -Xlinker -rpath -Xlinker /usr/lib/swift -Xlinker -headerpad_max_install_names  -swift-version 4  -Xfrontend -define-availability -Xfrontend 'SwiftStdlib 9999:macOS 9999, iOS 9999, watchOS 9999, tvOS 9999' -Xfrontend -define-availability -Xfrontend 'SwiftStdlib 5.0:macOS 10.14.4, iOS 12.2, watchOS 5.2, tvOS 12.2' -Xfrontend -define-availability -Xfrontend 'SwiftStdlib 5.1:macOS 10.15, iOS 13.0, watchOS 6.0, tvOS 13.0' -Xfrontend -define-availability -Xfrontend 'SwiftStdlib 5.2:macOS 10.15.4, iOS 13.4, watchOS 6.2, tvOS 13.4' -Xfrontend -define-availability -Xfrontend 'SwiftStdlib 5.3:macOS 11.0, iOS 14.0, watchOS 7.0, tvOS 14.0' -Xfrontend -define-availability -Xfrontend 'SwiftStdlib 5.4:macOS 11.3, iOS 14.5, watchOS 7.4, tvOS 14.5' -Xfrontend -define-availability -Xfrontend 'SwiftStdlib 5.5:macOS 12.0, iOS 15.0, watchOS 8.0, tvOS 15.0' -Xfrontend -define-availability -Xfrontend 'SwiftStdlib 5.6:macOS 12.3, iOS 15.4, watchOS 8.5, tvOS 15.4' -Xfrontend -define-availability -Xfrontend 'SwiftStdlib 5.7:macOS 13.0, iOS 16.0, watchOS 9.0, tvOS 16.0' -Xfrontend -define-availability -Xfrontend 'SwiftStdlib 5.8:macOS 9999, iOS 9999, watchOS 9999, tvOS 9999'  -F '/Users/ec2-user/jenkins/workspace/swift-PR-macos-smoke-test/branch-main/build/buildbot_incremental/swift-macosx-x86_64/lib' -Xlinker -rpath -Xlinker '/Users/ec2-user/jenkins/workspace/swift-PR-macos-smoke-test/branch-main/build/buildbot_incremental/swift-macosx-x86_64/lib' -module-cache-path /Users/ec2-user/jenkins/workspace/swift-PR-macos-smoke-test/branch-main/build/buildbot_incremental/swift-macosx-x86_64/swift-test-results/x86_64-apple-macosx10.9/clang-module-cache /Users/ec2-user/jenkins/workspace/swift-PR-macos-smoke-test/branch-main/swift/test/Interop/Cxx/stdlib/overlay/std-string-overlay.swift -Xfrontend -enable-experimental-cxx-interop -Xfrontend -validate-tbd-against-ir=none -o /Users/ec2-user/jenkins/workspace/swift-PR-macos-smoke-test/branch-main/build/buildbot_incremental/swift-macosx-x86_64/test-macosx-x86_64/Interop/Cxx/stdlib/overlay/Output/std-string-overlay.swift.tmp/a.out -module-name main  && /Users/ec2-user/jenkins/workspace/swift-PR-macos-smoke-test/branch-main/swift/utils/swift-darwin-postprocess.py /Users/ec2-user/jenkins/workspace/swift-PR-macos-smoke-test/branch-main/build/buildbot_incremental/swift-macosx-x86_64/test-macosx-x86_64/Interop/Cxx/stdlib/overlay/Output/std-string-overlay.swift.tmp/a.out && /usr/bin/env DYLD_LIBRARY_PATH='/Users/ec2-user/jenkins/workspace/swift-PR-macos-smoke-test/branch-main/build/buildbot_incremental/swift-macosx-x86_64/lib/swift/macosx'  /Users/ec2-user/jenkins/workspace/swift-PR-macos-smoke-test/branch-main/build/buildbot_incremental/swift-macosx-x86_64/test-macosx-x86_64/Interop/Cxx/stdlib/overlay/Output/std-string-overlay.swift.tmp/a.out
--
Exit Code: 134

Command Output (stdout):
--
[ RUN      ] std::string overlay.std::string <=> Swift.String
[       OK ] std::string overlay.std::string <=> Swift.String
[ RUN      ] std::string overlay.std::string as Swift.Sequence
stderr>>> generic witness table at 0x103a7c354 contains out-of-bounds requirement descriptor 0x103a7c80c
stderr>>> CRASHED: SIGABRT
the test crashed unexpectedly
[     FAIL ] std::string overlay.std::string as Swift.Sequence
std::string overlay: Some tests failed, aborting
UXPASS: []
FAIL: ["std::string as Swift.Sequence"]
SKIP: []
```

**Comment by egorzhdan** [comment]
@swift-ci please clean smoke test macOS

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

**Comment by egorzhdan** [comment]
@swift-ci please smoke test Linux

**Comment by egorzhdan** [comment]
@swift-ci please smoke test Linux

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

**Comment by egorzhdan** [comment]
Windows failure with `-v`:
```
```

File: lib/IRGen/GenDecl.cpp [hunk 1]
```diff
@@ -481,15 +481,25 @@ void IRGenModule::emitSourceFile(SourceFile &SF) {
...
       this->addLinkLibrary(LinkLibrary("c++", LibraryKind::Library));
-    else if (Context.LangOpts.Target.isOSLinux())
+    else if (target.isOSLinux())
       this->addLinkLibrary(LinkLibrary("stdc++", LibraryKind::Library));
 
     // Do not try to link Cxx with itself.
     if (!getSwiftModule()->getName().is("Cxx"))
       this->addLinkLibrary(LinkLibrary("swiftCxx", LibraryKind::Library));
+
+    // Only link with std on platforms where the overlay is available.
```
Comment (compnerd) [inline lib/IRGen/GenDecl.cpp:497]: Uh, the overlay is available on windows as well.  This seems wrong?

Comment (egorzhdan) [reply lib/IRGen/GenDecl.cpp:497]: `swiftstd` is not built on Windows, only `swiftCxx` is available currently.
I've tried enabling `swiftstd` on Windows (https://github.com/apple/swift/pull/62232) but ran into build issues that I haven't figured out so far.

File: stdlib/cmake/modules/SwiftSource.cmake [hunk 2]
```diff
@@ -372,7 +375,7 @@ function(_compile_swift_files
   cmake_parse_arguments(SWIFTFILE
     "IS_MAIN;IS_STDLIB;IS_STDLIB_CORE;IS_SDK_OVERLAY;EMBED_BITCODE;STATIC"
     "OUTPUT;MODULE_NAME;INSTALL_IN_COMPONENT;MACCATALYST_BUILD_FLAVOR;BOOTSTRAPPING"
-    "SOURCES;FLAGS;DEPENDS;SDK;ARCHITECTURE;OPT_FLAGS;MODULE_DIR"
+    "SOURCES;FLAGS;DEPENDS;SDK;ARCHITECTURE;OPT_FLAGS;MODULE_DIR;NO_LINK_NAME"
```
Comment (etcwilde) [inline stdlib/cmake/modules/SwiftSource.cmake:378]: Why is `NO_LINK_NAME` a list option? It doesn't look like it gets used in this function.

**Pull request merged**